acinclude/krb5.m4

```diff
@@ -8,125 +8,156 @@
-
+  AS_IF([test "x$squid_cv_broken_krb5_h" = "xyes"],[
+    AC_DEFINE(HAVE_BROKEN_SOLARIS_KRB5_H,1,[Define to 1 if Solaris krb5.h is broken for C++])
+  ])
+])
 
-AC_DEFUN([SQUID_CHECK_KRB5_HEIMDAL_BROKEN_KRB5_H], [
+AC_DEFUN([SQUID_CHECK_KRB5_HEIMDAL_BROKEN_KRB5_H],[
   AC_CACHE_CHECK([for broken Heimdal krb5.h],squid_cv_broken_heimdal_krb5_h, [
     SQUID_STATE_SAVE(squid_krb5_heimdal_test)
     CPPFLAGS="-I${srcdir:-.} $CPPFLAGS"
-    AC_RUN_IFELSE([AC_LANG_SOURCE([[
-#include <krb5.h>
-int
-main(void)
-{
-        krb5_context context;
-
-        krb5_init_context(&context);
-
-        return 0;
-}
-]])], [ squid_cv_broken_heimdal_krb5_h=no ], [
-    AC_RUN_IFELSE([AC_LANG_SOURCE([[
+    AC_RUN_IFELSE([
+      AC_LANG_PROGRAM([[#include <krb5.h>]],[[krb5_context c; krb5_init_context(&c);]])
+    ],[squid_cv_broken_heimdal_krb5_h=no],[
+      AC_RUN_IFELSE([
+        AC_LANG_PROGRAM([[
 #define HAVE_BROKEN_HEIMDAL_KRB5_H  1
 #include "compat/krb5.h"
-int
-main(void)
-{
-        krb5_context context;
+        ]],[[krb5_context c; krb5_init_context(&c);]])
+      ],[squid_cv_broken_heimdal_krb5_h=yes],[squid_cv_broken_heimdal_krb5_h=no])
+    ])
+    SQUID_STATE_ROLLBACK(squid_krb5_heimdal_test)
+  ])
+  AS_IF([test "x$squid_cv_broken_heimdal_krb5_h" = "xyes"],[
+    AC_DEFINE(HAVE_BROKEN_HEIMDAL_KRB5_H,1,[Define to 1 if Heimdal krb5.h is broken for C++])
+  ])
+])
 
-        krb5_init_context(&context);
+dnl checks that krb5 is functional. Sets squid_cv_working_krb5
+AC_DEFUN([SQUID_CHECK_WORKING_KRB5],[
+  AS_CASE("$1",
+    ["Heimdal"],[SQUID_CHECK_KRB5_HEIMDAL_BROKEN_KRB5_H],
+    ["Solaris"],[SQUID_CHECK_KRB5_SOLARIS_BROKEN_KRB5_H]
+  )
+  AC_CACHE_CHECK([for working krb5],squid_cv_working_krb5,[
+    SQUID_STATE_SAVE(squid_krb5_test)
+    CPPFLAGS="-I${srcdir:-.} $CPPFLAGS"
+    AC_RUN_IFELSE([
+      AC_LANG_PROGRAM([[#include "compat/krb5.h"]],[[krb5_context c; krb5_init_context(&c);]])
+    ],[squid_cv_working_krb5=yes],[squid_cv_working_krb5=no],[:])])
+  SQUID_STATE_ROLLBACK(squid_krb5_test)
+  AS_IF([test "x$squid_cv_working_krb5" = "xyes"],[
+    AC_DEFINE(HAVE_KRB5,1,[KRB5 support])
+  ],[test "x$squid_cv_working_krb5" = "xno" -a `echo $LIBS | grep -i -c "(-)L"` -gt 0],[
+    AC_MSG_NOTICE([Check Runtime library path !])
+  ])
+])
 
-        return 0;
-}
-]])], [ squid_cv_broken_heimdal_krb5_h=yes ], [ squid_cv_broken_heimdal_krb5_h=no ])
+dnl check how to access the max skew in the krb5 context
+AC_DEFUN([SQUID_CHECK_KRB5_GET_MAX_TIME_SKEW],[
+  AC_CACHE_CHECK([for max_skew in struct krb5_context],squid_cv_max_skew_context,[
+    SQUID_STATE_SAVE(squid_krb5_test)
+    CPPFLAGS="-I${srcdir:-.} $CPPFLAGS"
+    AC_CHECK_LIB(krb5,[krb5_get_max_time_skew],[
+      squid_cv_max_skew_context="yes"],[
+      AC_COMPILE_IFELSE([
+        AC_LANG_PROGRAM([[#include "compat/krb5.h"]],[[krb5_context kc; kc->max_skew = 1;]])
+      ],[squid_cv_max_skew_context=replace],[squid_cv_max_skew_context=no],[squid_cv_max_skew_context=no])
```

Comment:
IMO, it is not clear _why_ the existence of krb5_context::max_skew member implies that we should risk _replacing_ the standard krb5_get_max_time_skew() API function with our own same-name macro. Should not we use krb5_get_max_time_skew() if it is available? I suspect we should not replace standard API functions, but if we should, then we probably need to explain why (with a source code comment).

Reply:
After noticing the surrounding AC_CHECK_LIB(krb, X) call, I now suspect that the intent here was not "replace their X with ours" but "provide our X because they are missing it". If this is true, then I suggest:

* renaming HAVE_MAX_SKEW_IN_KRB5_CONTEXT with PROVIDE_KRB5_GET_MAX_TIME_SKEW because we set HAVE_MAX_SKEW_IN_KRB5_CONTEXT _not_ when krb5_context::max_skew exists but when that member exists _and_ we need to use that member to provide krb5_get_max_time_skew()
* replacing `squid_cv_max_skew_context` and its "replace" value with `squid_cv_krb5_get_max_time_skew` and "provide" value, to avoid the implication that we are _replacing_ something (and that we are replacing something like `krb5_context::max_skew`).

... where X is krb5_get_max_time_skew(), of course.

I do _not_ insist on any _specific_ words/changes for this change request, but I do think the current code is confusing and can be easily improved along the lines I outlined above.

Reply:
I have refactored so the code should be clearer that this is about *where* to get the skew value from. Please review the changes in https://github.com/squid-cache/squid/pull/1157/commits/5799bdfcd48d4f448769ee25af39e40529ef7b93 to see if they make more sense to you.

Reply:
Very nice rewrite, thank you.